### PR TITLE
feat: Token refresh hook

### PR DESCRIPTION
**Pull request by svrakitin**
- [x] Add `oauth2.refresh_token_hook` to configure an endpoint which will be called during `refresh_token` grant to retrieve updated token claims. Hook will be noop if not configured.
- [x] Add `AccessRequestHook` interface and `RefreshTokenHook` implementation. 

## Related issue(s)

#2570 

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](docs/docs).

## Further Comments

I added `github.com/hashicorp/go-cleanhttp` as a dependency, but I can avoid it if necessary.

### CODE REVIEW

File: docs/docs/advanced.md [hunk 1]
```diff
@@ -184,6 +184,49 @@ compatibility):
...
+
+```json
+{
+  "subject": "foo",
+  "granted_scopes": ["openid", "offline"],
+  "granted_audience": []
+}
+```
+
+Hook has to respond with updated session data for access and ID tokens:
```
Comment (aeneasr) [inline docs/docs/advanced.md:212]: Maybe clear up that the full "session" / extra will be replaced?

**Review by aeneasr — COMMENTED** [review]
This is looking great, I think a few more tests and a bit of config changes and we can merge this! Sorry for the late review :)

**Comment by svrakitin** [comment]
@aeneasr Thanks for the comments, I will update the PR this week.

File: spec/config.json [hunk 2]
```diff
@@ -675,6 +675,20 @@
           "examples": [
             "https://my-example.app/logout-successful"
           ]
+        },
+        "hooks": {
```
[the remark this answers is not in the document]
Comment (svrakitin) [reply spec/config.json:679]: Do you mean `urls.oauth2.refresh_token_hook` or `oauth2.refresh_token_hook`?

File: oauth2/oauth2_auth_code_test.go [hunk 3]
```diff
@@ -908,6 +910,63 @@ func TestAuthCodeWithMockStrategy(t *testing.T) {
 						require.NoError(t, json.Unmarshal(body, &refreshedToken))
 					})
 
+					t.Run("refreshing the token should call hook if configured", func(t *testing.T) {
+						if strat.d != "jwt" {
```
[the remark this answers is not in the document]
Comment (svrakitin) [reply oauth2/oauth2_auth_code_test.go:914]: It should, I can introspect the token I guess.

File: spec/config.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (aeneasr) [reply spec/config.json:679]:
```
oauth2.refresh_token_hook
```

:)

File: oauth2/hook.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+		session, ok := requester.GetSession().(*Session)
+		if !ok {
+			return nil
+		}
+
+		reqBody := TokenRefreshHookRequest{
+			Subject:         session.GetSubject(),
+			GrantedScopes:   requester.GetGrantedScopes(),
+			GrantedAudience: requester.GetGrantedAudience(),
```
[the remark this answers is not in the document]
Comment (svrakitin) [reply oauth2/hook.go:65]: I included `client_id` into request, not sure we need anything else here tbh.

File: oauth2/oauth2_auth_code_test.go [hunk 5]
```diff
@@ -908,6 +910,63 @@ func TestAuthCodeWithMockStrategy(t *testing.T) {
...
+							t.Skip()
+						}
+
+						hs := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (svrakitin) [reply oauth2/oauth2_auth_code_test.go:914]: Added more tests.

**Comment by svrakitin** [comment]
Updated PR. Some tests seem flaky.

**Comment by svrakitin** [comment]
@aeneasr Please review when you've got time, don't want to leave this hanging.

File: oauth2/hook.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aeneasr) [reply oauth2/hook.go:65]: Ok makes sense, we can also add fields in the future!

File: docs/docs/advanced.md [hunk 6]
```diff
@@ -184,6 +184,52 @@ compatibility):
 }
 ```
 
+#### Updating claims at token refresh
+
+Hydra can be configured to retrieve updated token claims from an endpoint at
+token refresh, which provides updated claims for a given subject and scopes.
+This is similar to accepting consent request, where the application provides the
+session data by calling Hydra Admin API.
+
```
Comment (aeneasr) [inline docs/docs/advanced.md:193]:
```suggestion

:::note

This endpoint is called _before_ any logic in Ory Hydra is executed. If the hook, for example, returns an error, the refresh token will remain unused!

::note

```

File: docs/docs/advanced.md [hunk 7]
```diff
@@ -184,6 +184,52 @@ compatibility):
...
+You can configure `oauth2.refresh_token_hook` config key:
```
Comment (aeneasr) [inline docs/docs/advanced.md:194]:
```suggestion
session data by calling Hydra Admin API.

:::note

This endpoint is called _before_ any logic in Ory Hydra is executed. If the hook, for example, returns an error, the refresh token will remain unused!

::note

You can configure `oauth2.refresh_token_hook` config key:
```

**Review by aeneasr — APPROVED** [review]
Awesome, thank you! 🎉

**Comment by alee792** [comment]
Apologies if this isn't the right place, but could this hook be extended to update consent/OIDC claims like ACR and AMR? This could support step-up or forced re-authn. 

For example, a user logs in with basic credentials and the identity provider's initial flow grants an `id_token` with an AC indicating 1FA. Later out-of-band, the identity provider solicits additional factors without going through the auth code grant again. When queried, the IDP can return those new ACR and AMR values.

If the RefreshHookResponse included these "non-Extra" fields, we could strengthen or modify the authenticator assurance level of the token's session. These fields are already delegated to the identity provider and taken at face value by Hydra.

**Comment by kszafran** [comment]
@svrakitin I just noticed that this feature was released in v1.10.7. I'm trying to understand -- does this hook get called also on the initial flow, when the caller gets the access token for the first time? If not, is there already a way to customize access token claims with hooks during the initial flow?

**Comment by svrakitin** [comment]
@kszafran You pass initial claims when you accept consent request. Hook is only triggered on refresh.

**Pull request merged**